controllers/openstackmachine_controller.go

```diff
@@ -430,6 +436,56 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 	return ctrl.Result{}, nil
 }
 
+func (r *OpenStackMachineReconciler) reconcileSecurityGroupsToInstance(logger logr.Logger, openStackCluster *infrav1.OpenStackCluster, machine *clusterv1.Machine, openStackMachine *infrav1.OpenStackMachine, instanceStatus *compute.InstanceStatus, computeService *compute.Service, networkingService *networking.Service) error {
+	logger.Info("Reconciling security groups to instance")
+
+	desiredSecurityGroups := []*infrav1.SecurityGroup{}
+
+	// These security groups are currently not managed by CAPO and provided by the user
+	// Later we'll have an API to manage them.
+	for _, sg := range openStackMachine.Spec.SecurityGroups {
+		if sg.ID == "" {
+			// lookup security group by name
+			securityGroup, err := networkingService.GetSecurityGroupByName(sg.Name)
```

Comment:
I'm not sure about this for a couple of reasons:
* We're making an OpenStack API call to determine if we need to update (i.e. in a 'hot path')
* The non-uniqueness of names in OpenStack

I like the way we handle these in the cluster controller for network and subnet resources: we 'resolve' resources specified by filter during initial reconciliation and write their UUID to the status. Subsequently we just refer to the status.

If we applied that here, we'd have a new list of security group uuids in the status which is separate to `AppliedSecurityGroups` which we would resolve *before* machine creation. This list would be as immutable as the machine spec, so we would not need to consider updates to it.

So:
```
if len(spec.SecurityGroups) != len(status.MachineSecurityGroups) {
  status.MachineSecurityGroups = resolve(spec.SecurityGroups)
}

... subsequently only ever use status.MachineSecurityGroups
```

In this way we would only look them up once, and we wouldn't break if the user did anything silly like renaming one or creating another one with the same name.

Reply:
> We're making an OpenStack API call to determine if we need to update (i.e. in a 'hot path')

I think the network/subnet/router implementation in the `openstackcluster_controller` [does it too](https://github.com/shiftstack/cluster-api-provider-openstack/blob/f9f1bae50f425eda850b145664f5d4352b437a92/controllers/openstackcluster_controller.go#L443-L444).

> we'd have a new list of security group uuids in the status which is separate to `AppliedSecurityGroupIDs` which we would resolve before machine creation. This list would be as immutable as the machine spec, so we would not need to consider updates to it.

Just to make sure I understand it 100%, this Status field would only be for the Security Groups specified via `OpenStackMachine.Spec.SecurityGroup` right? If yes, I agree that makes sense and I'll make the change.

> In this way we would only look them up once, and we wouldn't break if the user did anything silly like renaming one or creating another one with the same name.

Very good point.

> `if len(spec.SecurityGroups) != len(status.MachineSecurityGroups)`

Here I'm confused whether you want to compare the list of just their length. But I think you want to compare length indeed, since the `status.MachineSecurityGroups` is immutable (such as the `OpenStackMachine.Spec` too anyway).

Most of this comment is just making sure I understand what you have in mind correctly, thanks :pray: !